google/fingerprinters/web/scripts/updater/commuity/zabbix/update.sh

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env bash
+
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+source ../../common.sh
+
+SCRIPT_PATH="$(cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P)"
+# Root path to the web fingerprinter plugin.
+PROJECT_ROOT="$(cd -- "${SCRIPT_PATH}/../../../.." >/dev/null 2>&1 ; pwd -P)"
+# Path to the configurations for starting a live instance of Zabbix.
+ZABBIX_APP_PATH="${SCRIPT_PATH}/app"
+# Path to the temporary data holder.
+TMP_DATA="/tmp/zabbix_fingerprints"
+# Path to the local git repository for Zabbix codebase.
+GIT_REPO="${TMP_DATA}/repo"
+# Path to the directory of all the updated fingerprints data.
+FINGERPRINTS_PATH="${TMP_DATA}/fingerprints"
+# Json data of the final result.
+JSON_DATA="${FINGERPRINTS_PATH}/fingerprint.json"
+# Binary proto data of the final result.
+BIN_DATA="${FINGERPRINTS_PATH}/fingerprint.binproto"
+# Read all the versions to be fingerprinted.
+readarray -t ALL_VERSIONS < "${SCRIPT_PATH}/versions.txt"
+mkdir -p "${FINGERPRINTS_PATH}"
+
+startZabbix() {
+  local version="$1"
+  pushd "${ZABBIX_APP_PATH}" >/dev/null
+    ZABBIX_VERSION="${version}" docker-compose up -d
+  popd >/dev/null
+}
+
+stopZabbix() {
+  local version="$1"
+  pushd "${ZABBIX_APP_PATH}" >/dev/null
+    ZABBIX_VERSION="${version}" docker-compose down --volumes --remove-orphans
+  popd >/dev/null
+}
+
+# Convert the existing data file to a human-readable json file.
+convertFingerprint \
+  "${PROJECT_ROOT}/src/main/resources/fingerprinters/web/data/commuity/zabbix.binproto" \
+  "${JSON_DATA}"
+
+# Fetch Zabbix codebase.
+if [[ ! -d "${GIT_REPO}" ]] ; then
+  git clone https://github.com/zabbix/zabbix "${GIT_REPO}"
+fi
+
+# Update for all the versions listed in versions.txt file.
+for zabbix_version in "${ALL_VERSIONS[@]}"; do
+  echo "Fingerprinting Zabbix version ${zabbix_version} ..."
+  # Start a live instance of Zabbix.
+  startZabbix "${zabbix_version}"
+  # Arbitrarily chosen so that Zabbix is up and running.
+  echo "Waiting for Zabbix ${zabbix_version} to be ready ..."
+  sleep 30
+  TMP_VERSION=`echo ${zabbix_version}|grep -o '\d\.\d\.\d'`
+  echo ${TMP_VERSION}
+  # Checkout the repository to the correct tag.
+  checkOutRepo "${GIT_REPO}" "${TMP_VERSION}"
+
+  updateFingerprint \
```

Comment:
Could you regenerate the fingerprints after the updates made here? Your zabbix.binproto contains version 5.4.0, however there's no 5.4.0 in zabbix github tags. Make sure that before fingerprint regeneration, delete the zabbix.binproto in the resource folder, otherwise the updater would read in 5.4.0 again from the existing file and produce the incorrect results again.

There is a small bug on the updater side when init the fingerprint file for the first time (by passing in --init to :runFingerprintUpdater). `/tmp/zabbix_fingerprints/fingerprints/fingerprint.json` has to exist. I will make a bug fix to create a file if it doesn't exist.

Reply:
I still see [5.4.0] version returned when running Tsunami fingerprinter with the generated binary proto here, when running Zabbix 4.0.5. 

I will do more debugging by generating the binary proto using this script, will get back on this once I find the issue.

Reply:
It's pretty complicated to generate the initial json fingerprint file. You have to do the following:

1. Comment out 1st convertFingerprint in this file
2. Just have a single version in versions.txt
3. Pass in --init to runFingerprintUpdater in the common.sh

It should generate a single fingerprint.json.

Essentially call runFingerprintUpdater with --init to seed fingerprint.json

Then you can revert your script & common.sh as they should be, and iteratively go through all the versions.

I've only run through a few versions, the textproto should look like this: 
```
 {
    "contentPath": "conf/zabbix.conf.php.example",
    "versions": [{
      "fullName": "4.0.0"
    }, {
      "fullName": "4.0.1"
    }, {
      "fullName": "4.0.2"
    }, {
      "fullName": "4.0.3"
    }, {
      "fullName": "4.0.4"
    }, {
      "fullName": "4.0.5"
    }]
```
I will improve the documentation on my side, so it will be clearer how to do the fingerprint generation the first time.